### ISSUE TITLE
Event when a KV value changes

### ISSUE THREAD

**Issue by arpontes** [G-Research/consuldotnet#37]
Hello,

I'm using consuldotnet for a long time, but when I need to check if a value changes, I need to do it on my own. I think would be nice if this lib could do a check every X minutes for a change in the value and then fire and event with the new value.

I could implement this funcionality and send an PR, if you approve such feature.

**Comment by mfkl** [comment]
Hi,

Not sure that's in scope with consuldotnet, but maybe. Could you share a snippet of the solution you are currently using please?

> I'm using consuldotnet for a long time, but when I need to check if a value changes, I need to do it on my own.

Do you use `consul watch` for this or your own code?

If you suggest custom code for this, then users will probably ask next for a HTTP retry/error handling mechanisms, internet connection check, etc. I'm not sure this is in scope with this project, but let's see what others say.

Relevant:
- https://medium.com/@joatmon08/manage-monitor-consul-kv-changes-47e2a7a2d73d
- https://www.consul.io/commands/watch
- https://www.consul.io/docs/dynamic-app-config/watches

**Comment by lucascebertin** [comment]
Well, just trying to compare consul utilization across some languages and nodejs lib brings a `watch` functionality.
https://github.com/silas/node-consul#watch

Snippet:
```
import consulFn from 'consul'

const consul = consulFn()
const watch = consul.watch({
  method: consul.kv.get,
  options: { key: 'test' },
  backoffFactor: 1000,
});

watch.on('change', function(data, res) {
  console.log('data:', data);
});

watch.on('error', function(err) {
  console.log('error:', err);
});

setTimeout(function() { watch.end(); }, 30 * 1000);
```

@arpontes, any chance to this being related?
Could you share how you do that in dotnet today?

Thanks

**Comment by chester89** [comment]
I'd like watch functionality [for service instances](https://developer.hashicorp.com/consul/docs/dynamic-app-config/watches#service). Will the trick similar to KV work? e.g.
`var result = await consulClient.Health.Service("serviceName", "tagName", true, new QueryOptions
                {
                    WaitIndex = 0ul
                }, cancellationToken); //further processing`

**Comment by marcin-krystianc** [comment]
> I'd like watch functionality [for service instances](https://developer.hashicorp.com/consul/docs/dynamic-app-config/watches#service). Will the trick similar to KV work? e.g.
> `var result = await consulClient.Health.Service("serviceName", "tagName", true, new QueryOptions { WaitIndex = 0ul }, cancellationToken); //further processing`

Yes, I think it will work. Mind, that `WaitIndex` need to be set to the value from the previous query result.
